Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/DeviceStorage.kt

```diff
@@ -0,0 +1,117 @@
+package de.rki.coronawarnapp.storage
+
+import android.annotation.TargetApi
+import android.app.usage.StorageStatsManager
+import android.content.Context
+import android.os.Build
+import android.os.storage.StorageManager
+import android.text.format.Formatter
+import de.rki.coronawarnapp.util.ApiLevel
+import de.rki.coronawarnapp.util.storage.StatsFsProvider
+import timber.log.Timber
+import java.io.File
+import java.util.UUID
+
+// TODO Add inject when #1069 is merged
+class DeviceStorage constructor(
+    private val context: Context,
+    private val apiLevel: ApiLevel,
+    private val statsFsProvider: StatsFsProvider
+) {
+
+    private val privateStorage = context.filesDir
+    private val storageManager by lazy { context.getSystemService(Context.STORAGE_SERVICE) as StorageManager }
+
+    @TargetApi(Build.VERSION_CODES.O)
+    private fun requestStorageAPI26Plus(targetPath: File, requiredBytes: Long = -1L): CheckResult {
+        Timber.tag(TAG).v(
+            "requestStorageAPI26Plus(path=%s, requiredBytes=%d)", targetPath, requiredBytes
+
+        )
+        val statsManager =
+            context.getSystemService(Context.STORAGE_STATS_SERVICE) as StorageStatsManager
+
+        val storageUUID: UUID = storageManager.getUuidForPath(targetPath)
+
+        val totalBytes = statsManager.getTotalBytes(storageUUID)
+        var availableBytes = statsManager.getFreeBytes(storageUUID)
+
+        if (availableBytes < requiredBytes) {
+            val allocatableBytes = storageManager.getAllocatableBytes(storageUUID)
+            if (allocatableBytes + availableBytes >= requiredBytes) {
+                val toAllocate = requiredBytes - availableBytes
+                Timber.tag(TAG).v(
+                    "Not enough free space, allocating %d on %s.", requiredBytes, targetPath
+                )
+                storageManager.allocateBytes(storageUUID, toAllocate)
```

Comment:
This method may take several seconds to complete, so it should only be called from a worker thread.

Is this ensured by design?

Reply:
By design, this is not ensured by design :grin: , but must be done by the caller.
This will likely be called from the UI when making changes to options, or prior to downloading files.
We can't and don't want to make any assumptions here.

I'll clarify this in the entry method `checkSpacePrivateStorage` again. I think we can also add `@WorkerThread` to get some LINTer support with this.